### PR TITLE
Add httpupstream backend

### PR DESCRIPTION
**Pull request by freman**
Based slightly on a caddy module I just wrote to proxy authentication back to our internal gitlab installation.

### CODE REVIEW

File: httpupstream/auth_test.go [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+	defer ts.Close()
+	u, _ := url.Parse(ts.URL)
+
+	auth, err := NewAuth(u, time.Second, false)
+	NoError(t, err)
+
+	authenticated, err := auth.Authenticate("unknown", "secret")
+	NoError(t, err)
+	False(t, authenticated)
+}
```
Comment (smancke) [inline httpupstream/auth_test.go]: Should also contain _one_ positive test, here.

File: README.md [hunk 2]
```diff
@@ -197,6 +198,21 @@ Example:
...
 
+### Httpupstreem
+Authentication against an upstream http server by performing a simple simple authenticated request
+
+Parameters for the provider:
+| Parameter-Name    | Description                                            |
+| ------------------|--------------------------------------------------------|
+| upstream          | http/https url to call                                 |
+| skipverify        | true to ignore TLS errors (optional, false by default) |
+| timeout           | request timeout (optional 1m by default)               |
```
Comment (smancke) [inline README.md]: should mention, that the go duration syntax is supported.

File: httpupstream/backend.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+		return nil, errors.New(`missing parameter "upstream" for httpupstream provider`)
+	}
+
+	u, err := url.Parse(us)
+	if err != nil {
+		return nil, fmt.Errorf(`invalid parameter value "%s" in "upstream" httpupstream provider: %v`, us, err)
+	}
+
+	v := false
+	t := time.Minute
```
Comment (smancke) [inline httpupstream/backend.go]: could be extracted to a constant with a meaningful name, e.g. _defaultTimeout_

File: httpupstream/backend_test.go [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+
+	NoError(t, err)
+	Equal(t,
+		"https://google.com",
+		backend.(*Backend).auth.upstream.String())
+	Equal(t,
+		false,
+		backend.(*Backend).auth.skipverify)
+	Equal(t,
+		time.Second*60,
```
Comment (smancke) [inline httpupstream/backend_test.go]: should refer a constant, e.g. _defaultTimeout_ (see above commen)

File: README.md [hunk 5]
```diff
@@ -197,6 +198,21 @@ Example:
 loginsrv -backend 'provider=htpasswd,file=users
 ```
 
+### Httpupstreem
+Authentication against an upstream http server by performing a simple simple authenticated request
```
Comment (freman) [inline README.md]: lol that's what the second 'simple' should have been "simple basic authentication request" - I probably should have gone to bed before submitting this.

File: README.md [hunk 6]
```diff
@@ -26,6 +26,7 @@ The following providers (login backends) are supported.
 * [Htpasswd](#htpasswd)
 * [OSIAM](#osiam)
 * [Simple](#simple) (user/password pairs by configuration)
+* [Httpupstreem](#httpupstream)
```
Comment (freman) [inline README.md]: I typo like a pro

File: README.md [hunk 7]
```diff
@@ -197,6 +198,21 @@ Example:
 loginsrv -backend 'provider=htpasswd,file=users
 ```
 
+### Httpupstreem
```
Comment (freman) [inline README.md]: I typo consistently :D

File: README.md (hunk 2 above)
Comment (freman) [inline README.md]: Done

File: httpupstream/auth_test.go (hunk 1 above)
Comment (freman) [inline httpupstream/auth_test.go]: You now have two negatives and one positive tests :D

File: httpupstream/backend.go (hunk 3 above)
Comment (freman) [inline httpupstream/backend.go]: Done :D

File: httpupstream/backend_test.go (hunk 4 above)
Comment (freman) [inline httpupstream/backend_test.go]: see my above comment :D

**Comment by smancke** [comment]
I have added some test to get a full coverage and merged it.

Thanks for the contribution !!!

**Pull request merged**